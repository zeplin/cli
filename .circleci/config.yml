version: 2
jobs:
  build_and_test:
    docker:
    - image: circleci/node:10.18.0
    steps:
    - checkout
    # Restore node_modules if it is cached and has the exact same package-lock.json
    - restore_cache:
        key: dependency-cache-{{ checksum "package-lock.json" }}
        paths:
        - ./node_modules
    # Remove zeplin packages from node_modules because they might be updated with same tag since last caching
    - run:
        name: Remove zeplin packages from node_modules
        command: rm -rf node_modules/@zeplin*
    - run:
        name: Install npm packages
        command: npm install
    - run:
        name: Run linter
        command: npm run lint
    # Save node_modules to the cache for current package-lock.json
    - save_cache:
        key: dependency-cache-{{ checksum "package-lock.json" }}
        paths:
        - ./node_modules
    - run:
        name: Run tests
        command: npm test
  publish:
    docker:
    - image: circleci/node:10.18.0
    steps:
    - checkout
    - run:
        name: Publish to NPM
        command: |
          echo "@zeplin:registry=https://registry.npmjs.org/" >> ~/.npmrc
          echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
          npm publish --access public
workflows:
  version: 2
  npm_publish:
    jobs:
    - build_and_test:
        # Run build_and_test on all branches and tags
        filters:
          tags:
            only: /.*/
    - publish:
        requires:
        - build_and_test
        filters:
          # Ignore all branches
          branches:
            ignore: /.*/
          # Run publish only on version tags
          tags:
            only: /v[0-9]+(\.[0-9]+)*/
