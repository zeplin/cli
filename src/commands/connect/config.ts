import Joi from "@hapi/joi";
import * as fileUtil from "../../util/file";
import { ComponentConfigFile } from "./interfaces";
import { CLIError } from "../../errors";

const urlConfigSchema = Joi.object({
    type: Joi.string(),
    name: Joi.string().optional(),
    url: Joi.string()
});

const componentConfigSchema = Joi.object({
    path: Joi.string(),
    zeplinNames: Joi.array().items(Joi.string()).min(1),
    name: Joi.string().optional(),
    urlPaths: Joi.object().pattern(Joi.string(), Joi.string()).optional()
});

const componentConfigFileSchema = Joi.object({
    projects: Joi.array().items(Joi.string()).optional(),
    styleguides: Joi.array().items(Joi.string()).optional(),
    baseURLs: Joi.array().items(urlConfigSchema).optional(),
    components: Joi.array().items(componentConfigSchema).min(1)
}).custom((value, helpers) => {
    if (value.projects && value.styleguides) {
        if (value.projects.length === 0 && value.styleguides.length === 0) {
            throw new Error("at least one of `projects` or `styleguides` properties must contain 1 item.");
        }
    } else if (value.projects) {
        if (value.projects.length === 0) {
            throw new Error("`projects` must contain at least 1 item.");
        }
    } else if (value.styleguides) {
        if (value.styleguides.length === 0) {
            throw new Error("`styleguides` must contain at least 1 item.");
        }
    } else {
        return helpers.error("object.missing", { peersWithLabels: ["projects", "styleguides"] });
    }
    return value;
});

const getComponentConfigFile = async (filePath: string): Promise<ComponentConfigFile> => {
    const file = await fileUtil.readJsonFile(filePath);

    const { error, value } = componentConfigFileSchema.validate(file, { stripUnknown: true, presence: "required" });

    if (error) {
        throw new CLIError(`Oops! Looks like ${filePath} has some problems: ${error.message}`);
    }

    return value as ComponentConfigFile;
};

const getComponentConfigFiles = async (configFilePaths: string[]): Promise<ComponentConfigFile[]> => {
    try {
        const promises = configFilePaths.map(configFile => getComponentConfigFile(configFile));

        const configFiles = await Promise.all(promises);

        return configFiles;
    } catch (error) {
        throw new CLIError(error.message);
    }
};

export {
    getComponentConfigFile,
    getComponentConfigFiles
};
