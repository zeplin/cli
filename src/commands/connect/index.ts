export * from "./connect";
export * from "./delete";
export * from "./initialize";
