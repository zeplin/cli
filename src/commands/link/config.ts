import Joi from "@hapi/joi";
import * as fileUtil from "../../util/file";
import { ComponentConfigFile } from "./interfaces";
import { CLIError } from "../../errors";

const urlConfigSchema = Joi.object({
    type: Joi.string(),
    name: Joi.string().optional(),
    url: Joi.string()
});

const componentConfigSchema = Joi.object({
    path: Joi.string(),
    zeplinNames: Joi.array().items(Joi.string()).min(1),
    name: Joi.string().optional(),
    urlPaths: Joi.object().pattern(Joi.string(), Joi.string()).optional()
});

const componentConfigFileSchema = Joi.object({
    projects: Joi.array().items(Joi.string()).min(1).optional(),
    styleguides: Joi.array().items(Joi.string()).min(1).optional(),
    baseURLs: Joi.array().items(urlConfigSchema),
    components: Joi.array().items(componentConfigSchema).min(1)
}).or("projects", "styleguides");

const getComponentConfigFile = async (filePath: string): Promise<ComponentConfigFile> => {
    const file = await fileUtil.readJsonFile(filePath);

    const { error, value } = componentConfigFileSchema.validate(file, { stripUnknown: true, presence: "required" });

    if (error) {
        throw new CLIError(`Oops! Looks like ${filePath} has some problems: ${error.message}`);
    }

    return value as ComponentConfigFile;
};

const getComponentConfigFiles = async (configFilePaths: string[]): Promise<ComponentConfigFile[]> => {
    try {
        const promises = configFilePaths.map(configFile => getComponentConfigFile(configFile));

        const configFiles = await Promise.all(promises);

        return configFiles;
    } catch (error) {
        throw new CLIError(error.message);
    }
};

export {
    getComponentConfigFile,
    getComponentConfigFiles
};
