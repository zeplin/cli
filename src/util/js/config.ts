import * as fileUtil from "../file";

export interface PackageJson {
    [key: string]: object;
    dependencies: Record<string, string>;
    devDependencies: Record<string, string>;
    peerDependencies: Record<string, string>;
    scripts: Record<string, string>;
}

export async function getPackageJson(): Promise<PackageJson | null> {
    try {
        const packageJson = await fileUtil.readJsonFile("./package.json");

        return packageJson as PackageJson;
    } catch (e) {
        return null;
    }
}

export async function getBowerJson(): Promise<PackageJson | null> {
    try {
        const bowerJson = await fileUtil.readJsonFile("./bower.json");

        return bowerJson as PackageJson;
    } catch (e) {
        return null;
    }
}

export async function writePackageJson(packageJson: PackageJson): Promise<void> {
    await fileUtil.writeJsonIntoFile("./package.json", packageJson);
}
