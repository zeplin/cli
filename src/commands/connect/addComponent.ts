import chalk from "chalk";
import inquirer from "inquirer";
import dedent from "ts-dedent";
import {
    addComponentConfig,
    authentication,
    selectComponent,
    selectFile,
    selectResource,
    connectComponents
} from "../../tasks";
import { CliOptions, AddComponentContext } from "../../tasks/context/add-component";
import { Workflow } from "../../util/task";
import { indent } from "../../util/text";
import { AuthenticationService } from "../../service";
import { ConnectedComponentsService } from "./service";
import logger from "../../util/logger";
import { notInitialized, userSelectedNotToInitialize, addSummary } from "../../messages/initialize";
import { getComponentConfigFiles } from "./config";
import { initialize } from "..";
import { commandRunner } from "../../util/commander";

const askForInitialization = async (): Promise<boolean> => {
    const { confirmation } = await inquirer.prompt([{
        type: "confirm",
        message: "Do you want to initialize now?",
        default: false,
        name: "confirmation"
    }]);

    return confirmation;
};

export type AddComponentCommandOptions = CliOptions;

export async function addComponent(options: AddComponentCommandOptions): Promise<void> {
    try {
        const authService = new AuthenticationService();
        const connectService = new ConnectedComponentsService({ authService });

        const context: AddComponentContext = Object.assign(Object.create(null), {
            cliOptions: options
        }, {
            authService,
            connectService
        });

        const [existingConfigFile] = await getComponentConfigFiles([context.cliOptions.configFile]);
        if (!existingConfigFile) {
            logger.info(notInitialized());
            if (!(await askForInitialization())) {
                logger.info(userSelectedNotToInitialize());
                return;
            }

            return commandRunner(() => initialize(options))();
        }

        context.configFile = existingConfigFile;

        const workflow = new Workflow({
            context,
            tasks: [
                authentication,
                selectResource,
                selectComponent,
                selectFile,
                addComponentConfig,
                connectComponents
            ]
        });

        await workflow.run();

        logger.info(addSummary(context));
    } catch (error) {
        error.message = dedent`
            ${chalk.bold`Adding connected component failed.`}

            ${chalk.redBright(indent(error.message))}
        `;
        throw error;
    }
}
