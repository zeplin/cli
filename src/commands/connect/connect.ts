import chalk from "chalk";
import chokidar from "chokidar";
import dedent from "ts-dedent";
import logger from "../../util/logger";
import path from "path";
import { indent, stringify } from "../../util/text";
import { getComponentConfigFiles } from "./config";
import { ConnectedBarrelComponents } from "./interfaces/api";
import { connectComponentConfigFiles } from "./plugin";
import { ConnectDevServer } from "./server";
import { ConnectedComponentsService } from "./service";
import { isDefined } from "../../util/object";
import { flat } from "../../util/array";

const getComponentFilePaths = (connectedBarrels: ConnectedBarrelComponents[]): string[] => (
    flat(
        connectedBarrels.map(f =>
            f.items.map(c => (c.filePath ? path.resolve(c.filePath) : undefined)).filter(isDefined)
        )
    )
);

const generateConnectedComponents = async (
    options: Partial<Pick<ConnectOptions, "configFiles" | "plugins">>
): Promise<ConnectedBarrelComponents[]> => {
    const {
        configFiles,
        plugins
    } = options;

    const componentConfigFiles = await getComponentConfigFiles(configFiles, plugins);

    logger.debug(`component config files: ${stringify(componentConfigFiles)}`);

    const connectedBarrels = await connectComponentConfigFiles(componentConfigFiles);

    logger.debug(`connected barrels output: ${stringify(connectedBarrels)}`);

    return connectedBarrels;
};

const startDevServer = async (
    options: Pick<ConnectOptions, "configFiles" | "devModePort" | "devModeWatch" | "plugins">,
    connectedBarrels: ConnectedBarrelComponents[]
): Promise<void> => {
    const {
        configFiles,
        devModePort,
        devModeWatch,
        plugins
    } = options;

    logger.info("Starting development serverâ€¦");

    const devServer = new ConnectDevServer(connectedBarrels);

    if (devModeWatch) {
        let componentFiles = getComponentFilePaths(connectedBarrels);

        const watcher = chokidar.watch(
            [...configFiles, ...componentFiles],
            {
                cwd: process.cwd(),
                persistent: true,
                awaitWriteFinish: true
            }
        );

        watcher.on("change", async filePath => {
            logger.info((chalk.yellow(`\nFile change detected ${filePath}.\n`)));

            try {
                const updatedConnectedBarrels = await generateConnectedComponents({ configFiles, plugins });

                watcher.unwatch(componentFiles);

                devServer.updateConnectedBarrels(updatedConnectedBarrels);

                componentFiles = getComponentFilePaths(updatedConnectedBarrels);

                watcher.add(componentFiles);
            } catch (error) {
                logger.error(chalk.red(dedent`
                    Could not update Connected Components.
                    ${error}
                `));
            }
        });
    }

    await devServer.listen(devModePort);
};

const service = new ConnectedComponentsService();

const upload = async (connectedBarrels: ConnectedBarrelComponents[], { force }: ConnectOptions): Promise<void> => {
    logger.info("Connecting all Connected Components into Zeplinâ€¦");

    await service.uploadConnectedBarrels(connectedBarrels, { force });

    logger.info("ðŸ¦„ Components successfully connected to components in Zeplin.");
};

async function connect(options: ConnectOptions): Promise<void> {
    try {
        logger.debug(`connect options: ${stringify(options)}`);

        const connectedBarrels = await generateConnectedComponents(options);

        if (options.devMode) {
            await startDevServer(options, connectedBarrels);
        } else {
            await upload(connectedBarrels, options);
        }
    } catch (error) {
        error.message = dedent`
            ${chalk.bold`Connecting components to Zeplin components failed.`}

            ${chalk.redBright(indent(error.message))}
        `;
        throw error;
    }
}

export interface ConnectOptions {
    configFiles: string[];
    devMode: boolean;
    devModePort: number;
    devModeWatch: boolean;
    plugins: string[];
    force: boolean;
}

export {
    generateConnectedComponents,
    connect
};
