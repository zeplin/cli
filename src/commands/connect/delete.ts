import chalk from "chalk";
import dedent from "ts-dedent";
import { ConnectOptions } from ".";
import logger from "../../util/logger";
import { indent } from "../../util/text";
import { getComponentConfigFiles } from "./config";
import { ConnectedBarrels } from "./interfaces/api";
import { ConnectedComponentsService } from "./service";

const service = new ConnectedComponentsService();

const deleteConnectedBarrels = async (connectedBarrels: ConnectedBarrels[]): Promise<void> => {
    logger.info("Deleting Connected Components from Zeplinâ€¦");

    await service.deleteConnectedBarrels(connectedBarrels);

    logger.info("ðŸ”¥ Component connections successfully deleted from components in Zeplin.");
};

const getConnectedBarrels = async (options: ConnectDeleteOptions): Promise<ConnectedBarrels[]> => {
    const { configFiles } = options;

    const componentConfigFiles = await getComponentConfigFiles(configFiles, []);

    return componentConfigFiles.map(ccf => ({
        projects: ccf.projects || [],
        styleguides: ccf.styleguides || []
    }));
};

export type ConnectDeleteOptions = Pick<ConnectOptions, "configFiles">;

export async function connectDelete(options: ConnectDeleteOptions): Promise<void> {
    try {
        const barrels = await getConnectedBarrels({ configFiles: options.configFiles });

        await deleteConnectedBarrels(barrels);
    } catch (error: any) {
        error.message = dedent`
            ${chalk.bold`Deleting Connected Components from Zeplin components failed.`}

            ${chalk.redBright(indent(error.message))}
        `;
        throw error;
    }
}