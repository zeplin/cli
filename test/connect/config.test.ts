import path from "path";

import { getComponentConfigFiles } from "../../src/commands/connect/config";

const configResourcePath = path.resolve("test", "resources", "config");

describe("ConfigProcessor", () => {
    describe("when no config file option is passed", () => {
        describe("and no default config file is discovered", () => {
            it("fails", async () => {
                await expect(getComponentConfigFiles([], []))
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });
        });

        describe("and discovered config file is invalid", () => {
            it("fails", async () => {
                const searchFrom = path.join(configResourcePath, "invalid");

                await expect(getComponentConfigFiles([], [], searchFrom))
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });
        });

        describe("and json config file is discovered", () => {
            it("parses the config", async () => {
                const searchFrom = path.join(configResourcePath, "json");

                await expect(getComponentConfigFiles([], [], searchFrom))
                    .resolves
                    .toMatchSnapshot();
            });
        });

        describe("and json with comments config file is discovered", () => {
            it("parses the config", async () => {
                const searchFrom = path.join(configResourcePath, "jsonc");

                await expect(getComponentConfigFiles([], [], searchFrom))
                    .resolves
                    .toMatchSnapshot();
            });
        });

        describe("and yaml config file is discovered", () => {
            it("parses the config", async () => {
                const searchFrom = path.join(configResourcePath, "yaml");

                await expect(getComponentConfigFiles([], [], searchFrom))
                    .resolves
                    .toMatchSnapshot();
            });
        });
    });

    describe("when config file options are passed", () => {
        describe("and they are valid", () => {
            it("parses the config files", async () => {
                const configFilePath1 = path.join(configResourcePath, "json", ".zeplin", "components.json");
                const configFilePath2 = path.join(configResourcePath, "yaml", ".zeplin", "components.yaml");

                const configFilePaths = [configFilePath1, configFilePath2];

                await expect(getComponentConfigFiles(configFilePaths, []))
                    .resolves
                    .toMatchSnapshot();
            });
        });

        describe("and at least one is invalid", () => {
            it("fails", async () => {
                const configFilePath1 = path.join(configResourcePath, "json", ".zeplin", "components.json");
                const configFilePath2 = path.join(configResourcePath, "invalid", ".zeplin", "components.json");

                const configFilePaths = [configFilePath1, configFilePath2];

                await expect(getComponentConfigFiles(configFilePaths, []))
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });
        });
    });
});
