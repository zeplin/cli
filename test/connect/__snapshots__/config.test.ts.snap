// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProcessor when config file options are passed and at least one is invalid fails 1`] = `"Oops! Looks like test/resources/config/invalid/.zeplin/components.json has some problems: \\"components\\" must be an array"`;

exports[`ConfigProcessor when config file options are passed and they are valid parses the config files 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "path": "src/Sample1.jsx",
        "zeplinNames": Array [
          "screen1",
        ],
      },
      Object {
        "path": "src/Sample2.jsx",
        "zeplinNames": Array [
          "screen2",
        ],
      },
    ],
    "plugins": Array [],
    "projects": Array [
      "pid",
    ],
  },
  Object {
    "components": Array [
      Object {
        "path": "src/Sample1.jsx",
        "zeplinNames": Array [
          "screen1",
        ],
      },
      Object {
        "path": "src/Sample2.jsx",
        "zeplinNames": Array [
          "screen2",
        ],
      },
    ],
    "plugins": Array [],
    "projects": Array [
      "pid",
    ],
  },
]
`;

exports[`ConfigProcessor when no config file option is passed and discovered config file is invalid fails 1`] = `"Oops! Looks like test/resources/config/invalid/.zeplin/components.json has some problems: \\"components\\" must be an array"`;

exports[`ConfigProcessor when no config file option is passed and json config file is discovered parses the config 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "path": "src/Sample1.jsx",
        "zeplinNames": Array [
          "screen1",
        ],
      },
      Object {
        "path": "src/Sample2.jsx",
        "zeplinNames": Array [
          "screen2",
        ],
      },
    ],
    "plugins": Array [],
    "projects": Array [
      "pid",
    ],
  },
]
`;

exports[`ConfigProcessor when no config file option is passed and json with comments config file is discovered parses the config 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "path": "src/Sample1.jsx",
        "zeplinNames": Array [
          "screen1",
        ],
      },
      Object {
        "path": "src/Sample2.jsx",
        "zeplinNames": Array [
          "screen2",
        ],
      },
    ],
    "plugins": Array [],
    "projects": Array [
      "pid",
    ],
  },
]
`;

exports[`ConfigProcessor when no config file option is passed and no default config file is discovered fails 1`] = `
"Missing configuration file!
Please refer to [4mhttps://zpl.io/connected-components[24m to create a configuration file."
`;

exports[`ConfigProcessor when no config file option is passed and yaml config file is discovered parses the config 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "path": "src/Sample1.jsx",
        "zeplinNames": Array [
          "screen1",
        ],
      },
      Object {
        "path": "src/Sample2.jsx",
        "zeplinNames": Array [
          "screen2",
        ],
      },
    ],
    "plugins": Array [],
    "projects": Array [
      "pid",
    ],
  },
]
`;
